pkg/apis/monitoring/v1/types.go

```diff
@@ -41,6 +43,70 @@ type ByteSize string
 // +kubebuilder:validation:Pattern:="^(0|(([0-9]+)y)?(([0-9]+)w)?(([0-9]+)d)?(([0-9]+)h)?(([0-9]+)m)?(([0-9]+)s)?(([0-9]+)ms)?)$"
 type Duration string
 
+func isdigit(c byte) bool { return c >= '0' && c <= '9' }
+
+// Units are required to go in order from biggest to smallest.
+// This guards against confusion from "1m1d" being 1 minute + 1 day, not 1 month + 1 day.
+var unitMap = map[string]struct {
+	pos  int
+	mult uint64
+}{
+	"ms": {7, uint64(time.Millisecond)},
+	"s":  {6, uint64(time.Second)},
+	"m":  {5, uint64(time.Minute)},
+	"h":  {4, uint64(time.Hour)},
+	"d":  {3, uint64(24 * time.Hour)},
+	"w":  {2, uint64(7 * 24 * time.Hour)},
+	"y":  {1, uint64(365 * 24 * time.Hour)},
+}
+
+func (d Duration) AsTimeDuration() (time.Duration, error) {
```

Comment:
IIUC this is a copy from `github.com/prometheus/common/model`. Can't we call `model.ParseDuration()` in the controller?

I'm not a fan of additional functions to the api packages as it's easy to pull external dependencies by accident (https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#no-functions).

Reply:
Actually I wasn't aware of this function in common/model 😮 helps a lot!